Add RenderBox::lastLineBaseline and implementation for block flow objects

sammygill opened the pull request:
#### dc7fdb2e0a351c834cb5bd7e74eae44778e1cbcf
<pre>
Add RenderBox::lastLineBaseline and implementation for block flow objects
<a href="https://bugs.webkit.org/show_bug.cgi?id=245792">https://bugs.webkit.org/show_bug.cgi?id=245792</a>
rdar://100516446

Reviewed by Alan Bujtas.

In order to begin implementing last baseline alignment in various
sceanrios, we need to add a lastLineBaseline method in RenderBox similar
to how we have firstLineBaseline defined. This patch adds that method
along with an implementation for block items. Future patches that
implement last baseline alignment, such as for items in flex containers,
can use lastLineBaseline to get the last baseline of the items. Other
objects besides block objects will need an implementation as
necessary.

CSS Box Alignment Module Level 3 defines the last baseline of block
elements to be:  the dominant last baseline of the last in-flow line
box in the block container, or is taken from the last in-flow
block-level child in the block container that contributes a set of
last baselines, whichever comes first/last. The logic for this ends up
being very similar to the first baseline logic, but we are simply
starting from the last child and working our way up instead of the other
way around.

* Source/WebCore/layout/integration/inline/LayoutIntegrationLineLayout.cpp:
(WebCore::LayoutIntegration::LineLayout::firstLinePhysicalBaseline const):
(WebCore::LayoutIntegration::LineLayout::lastLinePhysicalBaseline const):
(WebCore::LayoutIntegration::LineLayout::physicalBaselineFromWritingMode const):
* Source/WebCore/layout/integration/inline/LayoutIntegrationLineLayout.h:
* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::firstLineBaseline const):
(WebCore::RenderBlock::lastLineBaseline const):
* Source/WebCore/rendering/RenderBlock.h:
* Source/WebCore/rendering/RenderBlockFlow.cpp:
(WebCore::RenderBlockFlow::lastLineBaseline const):
* Source/WebCore/rendering/RenderBlockFlow.h:
* Source/WebCore/rendering/RenderBox.h:
(WebCore::RenderBox::lastLineBaseline const):
(WebCore::RenderBox::firstInFlowChildBox const):
(WebCore::RenderBox::lastInFlowChildBox const):
(WebCore::RenderBox::nextInFlowSiblingBox const):

Canonical link: <a href="https://commits.webkit.org/255179@main">https://commits.webkit.org/255179@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f557c4a754c8747be4924ac2d44ee43a8dcf42a1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/91357 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/372 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/22011 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/101093 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/161055 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/95362 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/387 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/29320 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/83712 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/97472 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/97014 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/311 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/78120 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/27277 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/82236 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/81906 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/70305 "Updated gtk dependencies (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/35492 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/15932 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/33286 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/17018 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3594 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/37078 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/39839 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/38999 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/36127 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/layout/integration/inline/LayoutIntegrationLineLayout.cpp:
```diff
@@ -575,13 +576,29 @@ LayoutUnit LineLayout::firstLinePhysicalBaseline() const
+        return { };
+    }
+
+    Line& lastLine = m_inlineContent->lines.last();
```
auto would read better

sammygill replied:
Changed

alanbaradlay commented on Source/WebCore/layout/integration/inline/LayoutIntegrationLineLayout.cpp:
```diff
@@ -575,13 +576,29 @@ LayoutUnit LineLayout::firstLinePhysicalBaseline() const
+    return physicalBaselineFromWritingMode(lastLine);
+}
+
+LayoutUnit LineLayout::physicalBaselineFromWritingMode(LayoutIntegration::Line& line) const
```
physical (visual) values always incorporate the writing mode (hence the name). I'd just call it something like  physicalBaselineForLine.

sammygill replied:
Changed

alanbaradlay commented on Source/WebCore/rendering/RenderBlockFlow.cpp:
```diff
@@ -2998,6 +2999,32 @@ std::optional<LayoutUnit> RenderBlockFlow::firstLineBaseline() const
+        return std::nullopt;
+
+    const LayoutIntegration::LineLayout* lineLayout = modernLineLayout();
+    if (lineLayout)
```
if (auto* lineLayout = modernLineLayout()) ?

sammygill replied:
Changed

alanbaradlay commented on Source/WebCore/rendering/RenderBlockFlow.cpp:
```diff
@@ -2998,6 +2999,32 @@ std::optional<LayoutUnit> RenderBlockFlow::firstLineBaseline() const
+        return LayoutUnit { floorToInt(lineLayout->lastLinePhysicalBaseline()) };
+
+    ASSERT(lastRootBox()); 
+    LegacyRootInlineBox* rootBox = lastRootBox();
```
please use auto

sammygill replied:
Done!

alanbaradlay commented on Source/WebCore/rendering/RenderBox.h:
```diff
@@ -822,6 +826,11 @@ inline RenderBox* RenderBox::firstChildBox() const
 
+inline RenderBox* RenderBox::firstInFlowChildBox() const
+{
+    return downcast<RenderBox>(firstInFlowChild());
```
what guarantees that the first inflow child is always a RenderBox? (or you meant to call dynamicDowncast?)

sammygill replied:
Yes dynamicDowncast is what should have been used here. Changed.

alanbaradlay commented on Source/WebCore/rendering/RenderBox.h:
```diff
@@ -831,6 +840,11 @@ inline RenderBox* RenderBox::lastChildBox() const
 
+inline RenderBox* RenderBox::lastInFlowChildBox() const 
+{
+    return downcast<RenderBox>(lastInFlowChild());
```
same.

sammygill replied:
Changed

alanbaradlay commented on Source/WebCore/rendering/RenderBlock.cpp:
```diff
@@ -2546,13 +2546,25 @@ std::optional<LayoutUnit> RenderBlock::firstLineBaseline() const
-        }
+    for (RenderBox* child = firstInFlowChildBox(); child; child = child->nextInFlowSiblingBox()) {
+        if (auto baseline = child->firstLineBaseline())
+            return LayoutUnit { baseline.value() + child->logicalTop() };
```
I'd flip them just for clarity. top comes first and baseline is an offset from the top value.

sammygill replied:
Changed

sammygill commented on Source/WebCore/rendering/RenderBlock.cpp:
```diff
@@ -2546,13 +2546,25 @@ std::optional<LayoutUnit> RenderBlock::firstLineBaseline() const
+    if (isWritingModeRoot() && !isRubyRun())
+        return std::optional<LayoutUnit>();
+
+    for (RenderBox* child = lastInFlowChildBox(); child; child = child->nextInFlowSiblingBox()) {
```
Actually I'm pretty sure this is wrong. I'm pretty sure this should be child->previousInFlowSiblingBox() instead

sammygill committed:
Add RenderBox::lastLineBaseline and implementation for block flow objects
https://bugs.webkit.org/show_bug.cgi?id=245792
rdar://100516446

Reviewed by Alan Bujtas.

In order to begin implementing last baseline alignment in various
sceanrios, we need to add a lastLineBaseline method in RenderBox similar
to how we have firstLineBaseline defined. This patch adds that method
along with an implementation for block items. Future patches that
implement last baseline alignment, such as for items in flex containers,
can use lastLineBaseline to get the last baseline of the items. Other
objects besides block objects will need an implementation as
necessary.

CSS Box Alignment Module Level 3 defines the last baseline of block
elements to be:  the dominant last baseline of the last in-flow line
box in the block container, or is taken from the last in-flow
block-level child in the block container that contributes a set of
last baselines, whichever comes first/last. The logic for this ends up
being very similar to the first baseline logic, but we are simply
starting from the last child and working our way up instead of the other
way around.

* Source/WebCore/layout/integration/inline/LayoutIntegrationLineLayout.cpp:
(WebCore::LayoutIntegration::LineLayout::firstLinePhysicalBaseline const):
(WebCore::LayoutIntegration::LineLayout::lastLinePhysicalBaseline const):
(WebCore::LayoutIntegration::LineLayout::physicalBaselineFromWritingMode const):
* Source/WebCore/layout/integration/inline/LayoutIntegrationLineLayout.h:
* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::firstLineBaseline const):
(WebCore::RenderBlock::lastLineBaseline const):
* Source/WebCore/rendering/RenderBlock.h:
* Source/WebCore/rendering/RenderBlockFlow.cpp:
(WebCore::RenderBlockFlow::lastLineBaseline const):
* Source/WebCore/rendering/RenderBlockFlow.h:
* Source/WebCore/rendering/RenderBox.h:
(WebCore::RenderBox::lastLineBaseline const):
(WebCore::RenderBox::firstInFlowChildBox const):
(WebCore::RenderBox::lastInFlowChildBox const):
(WebCore::RenderBox::nextInFlowSiblingBox const):

Canonical link: https://commits.webkit.org/255179@main
